File: engines/cs/src/Engine.cs
using System;
using System.Linq;
using System.Text;

namespace SampleEngine
{
    class Engine
    {
        public bool ExitRequested { get; private set; }

        private readonly Action<string> m_writeLine;

        private Board? m_board = null;

        public Engine(Action<string> writeLine)
        {
            m_writeLine = writeLine;
        }

        public void Start()
        {
            Info();
        }

        public void ReadLine(string line)
        {
            var split = (line?.Trim() ?? "").Split(new char[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);

            var command = split.Length > 0 ? split[0] : "";
            var args = split.Length > 1 ? string.Join(" ", split, 1, split.Length - 1) : "";

            switch (command)
            {
                case Constants.CommandString_Info:
                    Info();
                    break;
                case Constants.CommandString_NewGame:
                    NewGame(args);
                    break;
                case Constants.CommandString_ValidMoves:
                    ValidMoves();
                    break;
                case Constants.CommandString_BestMove:
                    BestMove();
                    break;
                case Constants.CommandString_Play:
                    Play(args);
                    break;
                case Constants.CommandString_Pass:
                    Pass();
                    break;
                case Constants.CommandString_Undo:
                    Undo(args);
                    break;
                case Constants.CommandString_Options:
                    Options();
                    break;
                case Constants.CommandString_Exit:
                    Exit();
                    break;
                default:
                    WriteError(Constants.ErrorMessage_InvalidCommand);
                    break;
            }
        }

        private void WriteLine(string line)
        {
            m_writeLine(line);
        }

        private void WriteError(string message)
        {
            WriteLine($"{Constants.ErrString} {message}");
            WriteLine(Constants.OkString);
        }

        private void WriteError()
        {
            WriteError(Constants.ErrorMessage_Unknown);
        }

        private void Info()
        {
            WriteLine(Constants.IdString);
            WriteLine(Constants.OkString);
        }

        private void NewGame(string args)
        {
            m_board = new Board();

            string[] splitArgs = args.Split(new char[] { ';' }, StringSplitOptions.RemoveEmptyEntries);
            for (int i = 0; i < splitArgs.Length; i++)
            {
                if (i > 2)
                {
                    if (!m_board.TryParseMove(splitArgs[i], out Move? move, out string? moveString) || !m_board.TryPlayMove(move ?? Move.PassMove, moveString ?? ""))
                    {
                        WriteError(Constants.ErrorMessage_Unknown);
                        return;
                    }
                }
            }

            WriteLine(m_board.GetGameString());
            WriteLine(Constants.OkString);
        }

        private void ValidMoves()
        {
            if (m_board is null)
            {
                WriteError(Constants.ErrorMessage_NoGameInProgress);
                return;
            }

            if (m_board.GameIsOver)
            {
                WriteError(Constants.ErrorMessage_GameIsOver);
                return;
            }

            var validMoves = m_board.GetValidMoves();

            var sb = new StringBuilder();
            bool first = true;
            foreach (var validMove in validMoves)
            {
                if (m_board.TryGetMoveString(validMove, out string? result))
                {
                    sb.Append(first ? result : $";{result}");
                }
                first = false;
            }

            WriteLine(sb.ToString());
            WriteLine(Constants.OkString);
        }

        private void BestMove()
        {
            if (m_board is null)
            {
                WriteError(Constants.ErrorMessage_NoGameInProgress);
                return;
            }

            if (m_board.GameIsOver)
            {
                WriteError(Constants.ErrorMessage_GameIsOver);
                return;
            }

            var validMoves = m_board.GetValidMoves();

            if (validMoves.Count > 0 && m_board.TryGetMoveString(validMoves.First(), out string? result))
            {
                WriteLine(result ?? "");
                WriteLine(Constants.OkString);
            }
            else
            {
                WriteError();
            }
        }

        private void Play(string args)
        {
            if (m_board is null)
            {
                WriteError(Constants.ErrorMessage_NoGameInProgress);
                return;
            }

            if (m_board.GameIsOver)
            {
                WriteError(Constants.ErrorMessage_GameIsOver);
                return;
            }

            if (m_board.TryParseMove(args, out Move? move, out string? moveString) && m_board.TryPlayMove(move ?? Move.PassMove, moveString ?? ""))
            {
                WriteLine(m_board.GetGameString());
            }
            else
            {
                WriteLine($"{Constants.InvalidMoveString} {Constants.InvalidMoveMessage_Generic}");
            }

            WriteLine(Constants.OkString);
        }

        private void Pass()
        {
            Play(Constants.PassMoveString);
        }

        private void Undo(string args)
        {
            if (m_board is null)
            {
                WriteError(Constants.ErrorMessage_NoGameInProgress);
                return;
            }

            if (!int.TryParse(args, out int movesToUndo))
            {
                movesToUndo = 1;
            }

            if (movesToUndo < 1 || movesToUndo > m_board.CurrentTurn)
            {
                WriteError(Constants.ErrorMessage_UnableToUndo);
                return;
            }

            for (int i = 0; i < movesToUndo; i++)
            {
                m_board.TryUndoLastMove();
            }

            WriteLine(m_board.GetGameString());
            WriteLine(Constants.OkString);
        }

        private void Options()
        {
            WriteLine(Constants.OkString);
        }

        private void Exit()
        {
            ExitRequested = true;
        }
    }
}

File: engines/cpp/src/Engine.cpp
#include <cmath>
#include <sstream>

#include "Constants.h"
#include "Engine.h"

using namespace SampleEngine;

Engine::Engine(std::function<void(std::string)> writeLine) : m_writeLine{writeLine}
{
}

void Engine::Start()
{
    Info();
}

void Engine::ReadLine(std::string line)
{
    std::string command;
    std::string args;

    std::istringstream ss(line);
    std::ostringstream argStream;

    std::string token;

    int itemIndex = 0;
    while (std::getline(ss, token, ' '))
    {
        if (!token.empty())
        {
            if (itemIndex == 0)
            {
                command = token;
            }
            else if (itemIndex == 1)
            {
                argStream << token;
            }
            else
            {
                argStream << " " << token;
            }
            itemIndex++;
        }
    }

    args = argStream.str();

    if (command == CommandString_Info)
    {
        Info();
    }
    else if (command == CommandString_NewGame)
    {
        NewGame(args);
    }
    else if (command == CommandString_ValidMoves)
    {
        ValidMoves();
    }
    else if (command == CommandString_BestMove)
    {
        BestMove();
    }
    else if (command == CommandString_Play)
    {
        Play(args);
    }
    else if (command == CommandString_Pass)
    {
        Pass();
    }
    else if (command == CommandString_Undo)
    {
        Undo(args);
    }
    else if (command == CommandString_Options)
    {
        Options();
    }
    else if (command == CommandString_Exit)
    {
        Exit();
    }
    else
    {
        WriteError(ErrorMessage_InvalidCommand);
    }
}

void Engine::WriteLine(std::string line)
{
    m_writeLine(line);
}

void Engine::WriteError(std::string message)
{
    WriteLine(ErrString + (" " + message));
    WriteLine(OkString);
}

void Engine::WriteError()
{
    WriteError(ErrorMessage_Unknown);
}

void Engine::Info()
{
    WriteLine(IdString);
    WriteLine(OkString);
}

void Engine::NewGame(std::string args)
{
    m_board = std::make_shared<Board>();

    if (!args.empty())
    {
        std::istringstream ss(args);

        std::string token;

        int itemIndex = 0;
        while (std::getline(ss, token, ';'))
        {
            if (!token.empty())
            {
                if (itemIndex > 2)
                {
                    Move move;
                    std::string parsedMoveString;
                    if (!m_board->TryParseMove(token, move, parsedMoveString) ||
                        !m_board->TryPlayMove(move, parsedMoveString))
                    {
                        WriteError(ErrorMessage_Unknown);
                        return;
                    }
                }

                itemIndex++;
            }
        }
    }

    WriteLine(m_board->GetGameString());
    WriteLine(OkString);
}

void Engine::ValidMoves()
{
    if (!m_board)
    {
        WriteError(ErrorMessage_NoGameInProgress);
        return;
    }

    if (GameIsOver(m_board->GetBoardState()))
    {
        WriteError(ErrorMessage_GameIsOver);
        return;
    }

    auto validMoves = m_board->GetValidMoves();

    std::ostringstream str;
    bool first = true;
    for (auto const &iter : *validMoves)
    {
        std::string moveString;
        if (m_board->TryGetMoveString(iter, moveString))
        {
            if (!first)
            {
                str << ";";
            }
            str << moveString;
        }
        first = false;
    }

    WriteLine(str.str());
    WriteLine(OkString);
}

void Engine::BestMove()
{
    if (!m_board)
    {
        WriteError(ErrorMessage_NoGameInProgress);
        return;
    }

    if (GameIsOver(m_board->GetBoardState()))
    {
        WriteError(ErrorMessage_GameIsOver);
        return;
    }

    auto validMoves = m_board->GetValidMoves();

    Move bestMove = *(validMoves->begin());

    std::string result;
    if (m_board->TryGetMoveString(bestMove, result))
    {
        WriteLine(result);
        WriteLine(OkString);
    }
    else
    {
        WriteError();
    }
}

void Engine::Play(std::string args)
{
    if (!m_board)
    {
        WriteError(ErrorMessage_NoGameInProgress);
        return;
    }

    if (GameIsOver(m_board->GetBoardState()))
    {
        WriteError(ErrorMessage_GameIsOver);
        return;
    }

    Move move;
    std::string moveString;
    if (m_board->TryParseMove(args, move, moveString) && m_board->TryPlayMove(move, moveString))
    {
        WriteLine(m_board->GetGameString());
    }
    else
    {
        std::ostringstream invalidMove;
        invalidMove << InvalidMoveString << " " << InvalidMoveMessage_Generic;
        WriteLine(invalidMove.str());
    }

    WriteLine(OkString);
}

void Engine::Pass()
{
    Play(PassMoveString);
}

void Engine::Undo(std::string args)
{
    if (!m_board)
    {
        WriteError(ErrorMessage_NoGameInProgress);
        return;
    }

    std::istringstream ss(args);
    int movesToUndo;
    if ((ss >> movesToUndo).fail())
    {
        movesToUndo = 1;
    }

    if (movesToUndo < 1 || movesToUndo > m_board->GetCurrentTurn())
    {
        WriteError(ErrorMessage_UnableToUndo);
        return;
    }

    for (int i = 0; i < movesToUndo; i++)
    {
        m_board->TryUndoLastMove();
    }

    WriteLine(m_board->GetGameString());
    WriteLine(OkString);
}

void Engine::Options()
{
    WriteLine(OkString);
}

void Engine::Exit()
{
    m_exitRequested = true;
}
